### PR TITLE
port tdt and improve mendel errors, trio matrix

### PR DESCRIPTION
**Pull request by jbloom22**
Made relevant field and function names pythonic and consistent along the way.

I've ported the TDT python implementation and test quite literally. The implementation should use  sum on arrays but isn't yet properly exposed in api2.  I'm confused why -1 is used as a ploidy but will leave further improvements for later.

### CODE REVIEW

File: src/main/scala/is/hail/methods/MendelErrors.scala [hunk 1]
```diff
@@ -137,25 +138,25 @@ case class MendelErrors(hc: HailContext, vSig: Type, trios: IndexedSeq[CompleteT
 
   def mendelKT(): Table = {
     val signature = TStruct(
-      "fid" -> TString(),
-      "s" -> TString(),
+      "fam_id" -> TString(),
+      "id" -> TString(),
```
Comment (tpoterba) [inline src/main/scala/is/hail/methods/MendelErrors.scala]: why this change?

I think when we have general keys, we should just copy over the dataset col key field.

File: python/hail/methods/tests.py [hunk 2]
```diff
@@ -379,4 +379,29 @@ def test_export_plink(self):
...
\ No newline at end of file
+        # self.assertRaises(ExpressionException, lambda: methods.export_plink(ds, '/tmp/plink_example', id = ds.GT))
+        
+    def test_tdt(self):
+        pedigree = Pedigree.read('src/test/resources/tdt.fam')
+        tdt_tab = (methods.tdt(
+            methods.split_multi_hts(hc.import_vcf('src/test/resources/tdt.vcf', min_partitions=4)),
+            pedigree))
+
+        truth = hc.import_table('/Users/jbloom/hail/src/test/resources/tdt_results.tsv',
```
Comment (tpoterba) [inline python/hail/methods/tests.py]: relative path

File: python/hail/methods/family_methods.py [hunk 3]
```diff
@@ -26,22 +28,25 @@ def trio_matrix(dataset, pedigree, complete_trios=False):
 
     **Notes**
 
-    This method builds a new matrix table with one column per trio. If ``complete_trios`` is true,
-    then only trios that satisfy :meth:`~hail.representation.Trio.is_complete`
+    This method builds a new matrix table with one column per trio.
+    If ``complete_trios`` is true, then only trios that satisfy
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]:
```
If `complete_trios` is ``True``
```

File: python/hail/methods/family_methods.py [hunk 4]
```diff
@@ -26,22 +28,25 @@ def trio_matrix(dataset, pedigree, complete_trios=False):
...
+    :meth:`~hail.representation.Trio.is_complete`
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: This link is broken, just do `.Trio.is_complete`

File: python/hail/methods/family_methods.py [hunk 5]
```diff
@@ -26,22 +28,25 @@ def trio_matrix(dataset, pedigree, complete_trios=False):
...
+    are the sample IDs of the trio probands. The column fields and
     entries of the matrix are changed in the following ways:
 
-    The new column fields are:
+    The new column fields consist of three Structs (proband, father, mother),
+    a Boolean field, and a String field:
 
      - **proband.id** (*String*) - Proband sample ID, same as trio column key.
-     - **proband.annotations** (*Struct*) - Annotations on the proband.
+     - **proband.fields** (*Struct*) - Fields on the proband.
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: Column fields?

Comment (tpoterba) [reply python/hail/methods/family_methods.py]: This is inconsistent with the docs below, let's make them consistent somehow.

File: python/hail/methods/family_methods.py [hunk 6]
```diff
@@ -94,121 +99,107 @@ def mendel_errors(dataset, pedigree):
...
-    on more than one row. This table closely reflects the structure of the
-    ".mendel" PLINK format detailed below.
+    error, keyed by the variant and proband id.
 
     Columns:
 
-        - **fid** (*String*) -- Family ID.
-        - **s** (*String*) -- Proband ID.
+        - **fam_id** (*String*) -- Family ID.
+        - **id** (*String*) -- Proband ID. (key column)
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: column -> field

File: python/hail/methods/family_methods.py [hunk 7]
```diff
@@ -94,121 +99,107 @@ def mendel_errors(dataset, pedigree):
...
         - **v** (*Variant*) -- Variant in which the error was found.
+          (key column)
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: same

File: python/hail/methods/family_methods.py [hunk 8]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+    --------
+    Compute TDT association results and export to a file:
+
+    .. testsetup::
+    
+        tdt_dataset = hc.import_vcf('data/tdt_tiny.vcf')
+
+    >>> pedigree = Pedigree.read('data/tdt_trios.fam')
+    >>> tdt_table = methods.tdt(tdt_dataset, pedigree)
+    >>> tdt_table.export('output/tdt_results.tsv')
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: what if we `show` and put the output here rather than export? I think that's way clearer -- it reiterates the schema, field names, and that it's a table.

File: python/hail/methods/family_methods.py [hunk 9]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+
+    >>> tdt_table = tdt_table.filter(tdt_table.pval < 0.001)
+    >>> tdt_table.export("output/tdt_results.tsv")
+
+    Notes
+    -----
+    The
+    `transmission disequilibrium test <https://en.wikipedia.org/wiki/Transmission_disequilibrium_test#The_case_of_trios:_one_affected_child_per_family>`__
+    tracks the number of times the alternate allele is transmitted (t) or not
+    transmitted (u) from a heterozgyous parent to an affected child under the
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: unclear wording. It tracks the number of times the alternate allele is transmitted and compares to the null expectation of 0.5 transmissions per allele.

File: python/hail/methods/family_methods.py [hunk 10]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+    null that the rate of such transmissions is 0.5. For variants where
+    transmission is guaranteed (i.e., the Y chromosome, mitochondria, and
+    paternal chromosome X variants outside of the PAR), the test is invalid.
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: invalid or not performed? Should be clear about these variants being filtered out

Comment (tpoterba) [reply python/hail/methods/family_methods.py]: if they are

File: python/hail/methods/family_methods.py [hunk 11]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+    transmission is guaranteed (i.e., the Y chromosome, mitochondria, and
+    paternal chromosome X variants outside of the PAR), the test is invalid. 
+
+    The TDT statistic is given by
+
+    .. math::
+
+        (t-u)^2 \over (t+u)
+
+    and asymptotically follows a chi-squared distribution with 1 d.o.f.
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: write out degree of freedom

File: python/hail/methods/family_methods.py [hunk 12]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+         (    het,     het, hom_var,   auto, 0, 1),
+         (hom_var,     het,     het,   auto, 2, 0),
+         (hom_var,     het, hom_var,   auto, 1, 0),
+         (hom_var, hom_var,     het,   auto, 1, 0),
+         (hom_ref, hom_ref,     het, hemi_x, 0, 1),
+         (hom_ref, hom_var,     het, hemi_x, 0, 1),
+         (hom_var, hom_ref,     het, hemi_x, 1, 0),
+         (hom_var, hom_var,     het, hemi_x, 1, 0)]
+    
+    mapping = {functions.capture([v[0], v[1], v[2], v[3]]): [v[4], v[5]] for v in l}
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: can we make these structs? I think it'll be much clearer.

```
mapping = {Struct(kid=v[0], dad=v[1], mom=v[2], copy_state=v[3])): Struct(t=v[4], u=v[5]) for v in l} 
```

File: python/hail/methods/family_methods.py [hunk 13]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+                            tri.mother_entry.GT.num_alt_alleles(),
+                            ploidy])[0]))))
+    
+    tri = tri.annotate_rows(u = agg.sum(
+        agg.filter((tri.category != 1) | ~tri.father_entry.GT.is_het(),
+            functions.bind(
+                functions.cond(
+                    tri.category == 0,
+                    2,
+                    functions.cond(
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: hah, we definitely need a switch function.

File: python/hail/methods/family_methods.py [hunk 14]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+    tri = tri.annotate_globals(mapping = mapping)
+
+    tri = tri.annotate_rows(category =
+        functions.cond(tri.v.in_autosome_or_par(),
+                        0,
+                        functions.cond(tri.v.in_x_nonpar(),
+                                       1,
+                                       -1)))
+    
+    # FIXME calculate in one pass once aggs are re-vectorized
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: ? You should do this now, I don't see anywhere in the second annotate_rows that uses `tri.t`

File: python/hail/methods/family_methods.py [hunk 15]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+    
+    tab = tab.annotate(chi2 =
+        functions.cond(
+            tab.t + tab.u > 0,
+            ((tab.t - tab.u) ** 2) / (tab.t + tab.u),
+            0.0))
+    
+    tab = tab.annotate(pval = functions.pchisqtail(tab.chi2, 1.0))
+    
+    return tab
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: we should probably cache/persist this result. Thoughts?

File: python/hail/methods/family_methods.py [hunk 16]
```diff
@@ -225,4 +216,207 @@ def mendel_errors(dataset, pedigree):
...
+                        -1,
+                        functions.cond(
+                                tri.is_female,
+                                2,
+                                1))),
+                lambda ploidy: tri.mapping.get([
+                            tri.proband_entry.GT.num_alt_alleles(),
+                            tri.father_entry.GT.num_alt_alleles(),
+                            tri.mother_entry.GT.num_alt_alleles(),
+                            ploidy])[1]))))
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: ploidy is only used once here? If so, the bind is just unnecessary complexity.

Comment (tpoterba) [reply python/hail/methods/family_methods.py]: you also don't have to do all these computations in one python statement.

File: python/hail/methods/tests.py [hunk 17]
```diff
@@ -379,4 +379,29 @@ def test_export_plink(self):
         
         # FIXME still resolving: these should throw an error due to unexpected row / entry indicies, still looking into why a more cryptic error is being thrown
         # self.assertRaises(ExpressionException, lambda: methods.export_plink(ds, '/tmp/plink_example', id = ds.v.contig))
-        # self.assertRaises(ExpressionException, lambda: methods.export_plink(ds, '/tmp/plink_example', id = ds.GT))
\ No newline at end of file
+        # self.assertRaises(ExpressionException, lambda: methods.export_plink(ds, '/tmp/plink_example', id = ds.GT))
+        
+    def test_tdt(self):
+        pedigree = Pedigree.read('src/test/resources/tdt.fam')
```
Comment (tpoterba) [inline python/hail/methods/tests.py]: use `test_file('tdt.fam')` and below now that the recent PR is merged

**Comment by jbloom22** [comment]
I've gut-renovated the TDT implementation and improved docs and tests. The t=0, u=0 case is now handed generically, giving NaN and NaN instead of 0.0 and 1.0 for chi2 and pval. I've also normalized the pedigree and trio fields/functions to be consistent with the other changes.

File: python/hail/genetics/pedigree.py [hunk 18]
```diff
@@ -6,40 +6,40 @@
...
     :type is_female: bool or None
     """
 
     @handle_py4j
     @record_init
-    @typecheck_method(proband=strlike,
-                      fam=nullable(strlike),
-                      father=nullable(strlike),
-                      mother=nullable(strlike),
+    @typecheck_method(s=strlike,
```
Comment (tpoterba) [inline python/hail/genetics/pedigree.py]: I think this is wrong -- we should rename `s` to `pro_id` or something for consistency

File: python/hail/genetics/pedigree.py [hunk 19]
```diff
@@ -150,8 +152,8 @@ def is_female(self):
     def is_complete(self):
         """Returns True if the trio has a defined mother, father, and sex.
 
-        The considered fields are ``mother``, ``father``, and ``sex``.
-        Recall that ``proband`` may never be missing. The ``fam`` field
+        The considered fields are ``mat_id``, ``pat_id``, and ``is_female``.
```
Comment (tpoterba) [inline python/hail/genetics/pedigree.py]: use `:meth:` references here.
